### ISSUE TITLE
`ksh` doesn't use threads so what should the stance be vis-a-vis threads in plugins?

### ISSUE THREAD

**Issue by krader1961** [att/ast#357]
While resolving issue #276 I had to disable use of the `__thread` directive to make a var use thread local storage on OpenBSD 6.2. Obviously that is a hack that shouldn't be needed. But it raises the question of whether threads should be explicitly allowed or disallowed in dynamically loaded builtins.

**Comment by siteshwar** [comment]
Is it even possible to safely use threads in a shell builtin ? My understanding is that it's a bad idea to mix forks with threads and it should be disallowed.

**Comment by DavidMorano** [comment]
@siteshwar
> Is it even possible to safely use threads in a shell builtin ? 
> My understanding is that it's a bad idea to mix forks 
> with threads and it should be disallowed.

I have many KSH built-ins that are multithreaded. Further, and much more daring, I have several threads (some started dynamically) that actually run along side KSH even when no built-ins are running. These can best be described as daemon-server threads that capture various events outside of particular built-in use. The built-ins also use general purpose libraries (shared-object or otherwise) that have background threads of their own (mostly started dynamically). So at any one time during a KSH session, there may be potentially several or many threads all running within the KSH address space.

Of course, through all of this KSH does not know about the other threads.  Any interaction with the KSH API interfaces is serialized since KSH and its programming interface were not meant to be multithread safe. About forks (which of course KSH does all of the time): there is no problem at all with regular threads inside of a KSH built-in (assuming POSIX |fork1(2)| semantics, which should be arranged for if not the default already). But with background daemon threads, or with library background threads, there is the fork-safety problem. In general, all background threads need to be made "fork-safe" so that the child does not get messed up by half-baked variables in the address space that are no longer coherent due to an asynchronous fork by KSH itself. Also, it is always a *very* good idea to make sure that all background threads are eventually "joined" somehow.  This is generally done by arranging for the join to happen when the module that the background thread code is located in gets unloaded (|dlclose(3dl)|).  How this is done when the background threads are located in dynamically loaded and unloaded code modules is not obvious and is beyond the scope of this discussion. The UNIX thread-gods did not foresee this possibility and they did not plan for it specifically, but it is still achievable with some custom well-placed coding.

Anyway, the short of it is that multi-threaded built-ins are not only possible, but background threads, being background to the rest of KSH itself, is also possible with due care.  Some caveats: the AST VMALLOC code had to be compiled out.  Although it was supposed to have a "thread safe" mode, that did not turn out to be thread-safe enough for any amount of multi-threading at all -- as far as I experienced. Without VMALLOC, the default allocation facility reverts to that of the host OS system itself (last I looked), which is generally (as far as I know) always multi-thread safe. Also, if you want multi-thread SFIO (like from within a built-in), make sure that your SFIO is complied as multi-thread safe also. For myself, I do not use SFIO for my multi-threaded I/O. Rather, I use my own interface library for that and most everything else. But remember that the KSH built-in programming interface *requires* the use of SFIO for STDIN, STDOUT, and STDERR regardless of anything else one might do with I/O from within a built-in.

I have written over 110 KSH built-ins.  Several of them are explicitly multi-threaded.  But even my built-ins that are not explicitly multi-threaded, call library calls or instantiate library objects that may themselves be multi-threaded. So at any one time there can be zero or about a hand full (or more) background threads running along side and with KSH (without KSH explicitly knowing that they are even there). With properly written code, none of this is out of the ordinary.

In my view, all code today might secretly be multi-threaded for the simple reason that when libraries (either statically linked or more commonly dynamically linked) are called from any code (like from KSH built-in code) one or more background threads could be spawned.  It is just the way of the modern era now-a-days. People should just get used to this idea.

Main caveat for multi-thread use w/ KSH (as far as I know): compile *out* the VMALLOC stuff and use the host OS process heap management instead.

**Comment by fpmurphy** [comment]
@krader1961.   You are misinformed.  I concur with @DavidMoreno.   I too have written  a number of  threaded ksh93 plugins

**Comment by krader1961** [comment]
The issue is that to get the code to build on OpenBSD I had to disable

> ... the use of __thread from src/lib/libast/aso/asothreadid.c.

My question has nothing to do with the creation of threads by library functions that might be used by `ksh`. It has to do with writing plugins that depend (if only indirectly) on the AST "aso" code. So, @fpmurphy, I am not misinformed. My question was "whether threads should be explicitly allowed or disallowed in dynamically loaded builtins." Not whether fork/exec of external commands should work if non-main threads have been created. You're really chapping my ass with your uncharitable interpretations of everything I write. Note too that even @DavidMorano notes that the existing AST code may not work correctly in the face of multiple threads of execution.

Having said that: If plugins people are writing do not depend, even indirectly, on the AST "aso" code then my concern is a non-issue. If they do depend on that code then we need to fix the thread support on OpenBSD.

**Comment by DavidMorano** [comment]
Just to clarify from my experience with threads inside of plugins (or just running in the background of KSH):

> Having said that: If plugins people are writing do not depend, even indirectly, on the AST "aso" code then my concern is a non-issue. If they do depend on that code then we need to fix the thread support on OpenBSD.

In my experience, threads did not depend on anything specific within the AST except for the following:

+ had to disable the use of VMALLOC entirely (something in even the multithread-safe version of VMALLOC caused use of threads to crash); other people's experience may differ, and I would be happy to hear about it

+ although I did not use SFIO in a multithreaded way, any user who tries to do so SHOULD enable the multithreaded version of SFIO; reminder: why do we even care about SFIO? - because use of it (SFIO) for STDIN, SDTOUT,and STDERR is currently *required* by the KSH built-in plugin API

+ no LIBAST ASO is required for multithreaded operation; my view is that LIBAST ASO is provided for fun and is in no way needed to support user threads in plugins or otherwise; I did not check, but there might be a remote chance that LIBAST ASO is used in the multithreaded version of SFIO, but I doubt it (but someone might want to check on that)

+ regarding that thread-local storage qualifier from LIBAST ASO ('src/lib/libast/aso/asothradid.x')  '__thead' on the OpenBSD environment, that would seem to indicate that the OpenBSD compilation environment does not support that language element (it would seem); so LIBAST ASO cannot by built properly on an OpenBSD system -- so be it; in my opinion ASO is not (as far as I know) and should not be a requirement for either the compilation of the base KSH itself or for successful use of threads within KSH (either in built-ins or otherwise background)

At the risk of being too pedantic, more clarification from my experience using threads w/ KSH:

> My question ...  It has to do with writing plugins that depend (if only indirectly) on the AST "aso" code.

My threads do *not* depend on any LIBAST ASO code and I would recommend in general that other people not rely on LIBAST ASO being present in any KSH distros.  If someone wants to use LIBAST ASO, then that is their own decision and *not* -- in my view --  the responsibility of the KSH distro to provide it! Users wanting to use ASO should arrange for linking against that themselves, and not reply on it already being complied into the base KSH -- in my opinion. The only possible caveat to my statement here is that there is a remote possibility that a compile-option of a multithread-safe SFIO *might* depend on LIBAST ASO, but I doubt it. I only mention this because I have not investigated (or do not remember) the precise requirements for a successful compilation of a multithread-safe SFIO (which the SFIO developers optionally coded for). Just an additional clarification: SFIO was designed to optionally be multithread-safe but *not* multithreaded in itself.  I have not investigated, but this is my present understanding from memory.

Just to clarify further -- and from my particular experience:

> My question was "whether threads should be explicitly allowed or disallowed in dynamically loaded builtins." 

answer: ALLOWED
But other than my points raised above (disable VMALOC and optionally enable multithread-safe compile mode of SFIO), nothing else within the whole of the KSH distro should need to be included. Specifically, I am not aware that LIBAST ASO is required at all, but someone should just confirm that a multithread-safe compile of SFIO does not depend on LIBAST ASO. If one does not care about a multithread-safe version of SFIO (which includes myself), then even that is not an issue for a successful compile of KSH and the successful use of multithreaded plugins (or other random threads alongside KSH).

Additionally, for those adventurous enough to use threads in the background of KSH itself:
Do threads need to be made fork-safe? Answer: YES
Why? Because KSH forks a lot for its living. As everyone experienced with threads should already know, forks followed essentially "closely"  by execs (somewhat open to interpretation; see discussion on vfork for more dangers in this area) are not the problem (at least not in a POSIX compliant environment, which I assume the whole of this audience is in).  But KSH does a lot of other forks.  These other forks can be a problem for background threads.  Any threads in the background of KSH are going to experience a lot of forking, and so any background threads should be made fork-safe as may be necessary. But the truth of the matter is that this issue is in *no* way peculiar to KSH.  ANY threads existing in an environment where a fork might happen along side any running thread should be made fork-safe. In my view, any and all background or library threads should be made fork-safe. Why? Because there is always the likelihood that two things can always happen simultaneously in a modern program:
a) forks happening pseudo randomly, possibly being initiated from library calls (including standard LIBC library calls)
b) background threads may be initiated from library calls, otherwise unknown to the caller
A general purpose developer should *not* be required to know whether some library call initiates either a fork or a thread. Therefore -- and in my opinion -- all background or library-call initiated background threads should be made fork-safe. But none of this is peculiar to KSH itself.  This has everything to do with normal healthy general purpose library programming where forks and threads can exist happily side by side. In short. the use of forks and threads simultaneously is a topic in its own right and not peculiar at all to use within KSH. KSH is just an environment where you are guaranteed to get forks. so a threading programmer just has to deal with that (as they should anyway in my opinion).

Finally, and in the interest of full disclosure: are there "issues" with the use of successful (non-crashing) threads in code modules that can be dynamically unloaded at run-time (like with a |dlcose(3dl)| call? Yes, there are; serious issues (background threads located in code that gets asynchronously unloaded will cause a crash). Also registered call-backs located in code that can be unloaded will also cause a crash when that code gets unloaded and the call-back gets subsequently called. How these issues are properly handled is all a bit outside the scope of the issue of KSH itself though. This (again) is not peculiar to the use of threads in KSH per se.  It is a problem with the use of threads in dynamically loaded and unloaded shared object modules in general.  Thread related issues are, in general, not related per se w/ KSH. Rather, thread related issues are related to the general use of threads in any general purpose program environment.  KSH is just a general purpose (anything can happen) program environment.

Sorry for the length of this post, but I hope that this was at least helpful for those wondering about thread use within KSH. But this is not a lesson in how to use threads in hostile environments (like KSH is).  Using threads in a serious general purpose programming environment (where anything can and will happen) is additionally a topic in itself, outside the scope of KSH per se.

**Comment by krader1961** [comment]
Also, FWIW, there is at least one place where `ksh` does not use a threadsafe function. From `cppcheck`:

```
[../src/cmd/ksh93/bltins/cd_pwd.c:134] portability (getpwuidCalled):
 Non reentrant function 'getpwuid' called. For threadsafe applications it is recommended to use the reentrant replacement function 'getpwuid_r'.
```

So if we're going to officially support threads in plugins we need to address such issues.

>  in my opinion ASO is not (as far as I know) and should not be a requirement for either the compilation of the base KSH itself or for successful use of threads within KSH (either in built-ins or otherwise background)

Thank you, @DavidMorano, for that extremely thorough comment from which the above snippet is extracted. I'm especially intrigued that the ASO subsystem may not be needed to build `ksh` or the AST subsystems it depends on. I'll investigate if that is true. And if it isn't needed we'll rip it out.

**Comment by krader1961** [comment]
There are a handful of uses of the ASO code:

```
$ grep 'aso\.h' (ff)                                                3 ms
cmd/ksh93/include/jobs.h:#include <aso.h>
cmd/ksh93/sh/jobs.c:#include "aso.h"
cmd/ksh93/sh/fault.c:#include "aso.h"
cmd/tests/terror.h:#include <aso.h>
lib/libast/man/aso.3:#include <aso.h>
lib/libast/misc/debug.c:#include <aso.h>
lib/libast/include/aso.h:#include <ast_aso.h>
lib/libast/cdt/cdtlib.h:#include <aso.h>
lib/libast/vmalloc/vmhdr.h:#include <aso.h>    /* atomic scalar operations              */
lib/libast/aso/asohdr.h:#include "aso.h"
lib/libast/aso/aso.c:#include "ast_aso.h"
lib/libast/string/fmtbuf.c:#include <aso.h>
lib/libast/port/intercept.c:#include <aso.h>
```

I'll bet all of those can be removed or replaced by use of the `volatile` keyword or pthread alternatives. For example, `aso_yield()` (used in  *src/cmd/ksh93/sh/jobs.c*) is a trivial wrapper around `sched_yield()` that is part of POSIX pthreads support. So this seems like another example where this code made sense 20+ years ago but can be replaced by modern alternatives  that should be available on every platform we care about.

**Comment by DavidMorano** [comment]
@krader1961 
>>  Non reentrant function 'getpwuid' called.

> So if we're going to officially support threads in plugins we need to address such issues.

Yes, that should be cleaned up for its own sake, but having KSH use all reentrant calls is not a strictly necessary requirement for successful multithread use within either built-ins or background threads alongside KSH.  Contrary to most belief, non-thread-safe calls *can* be used in multithread programs -- with the important proviso that only a single thread is using the non-thread-safe calls.  This is true for all OSes that I am aware of. It is common, but not strictly necessary, that the "main" thread is the one thread normally allowed to make the non-thread-safe calls.  This just happens to be the case with KSH.  KSH forms the main thread, and all other threads are secondary.  As long as only one thread in a whole program is using the non-thread-safe calls, the rest of the program functions completely normally, assuming that it only uses thread-safe calls in all other threads and with some additional possible caveats mentioned later in this post (see |getpwxxx()| and friends below).

Although KSH uses the non-thread-safe library call you found above, there are other important problems within the general KSH/LIBAST environment.  Very much of -- if not perhaps most of -- the entire AST library calls are non-thread-safe!  It is just the way they were all originally coded.  This necessitates that most all AST calls cannot be used simultaneously from within multithreaded built-ins or from any background threads at all.  No LIBAST calls (or any other general AST library calls) can *ever* be made from a background thread, since this usage can always conflict with a simultaneous usage by KSH itself. What is possible (with some overhead care) is that serialized access to non-thread-safe LIBAST calls should be able to be made from the built-in threads (but not from background threads as already mentioned). The basic thing to remember is that built-in threads (although running simultaneous with each other) only run alternately with KSH itself.  But any backgrounds threads are (obviously) always running simultaneous with everything else. This arrangement forms the basic restrictions on what is possible and what is not.

One should also remember that any calls back into KSH itself are also not thread-safe.  These sorts of calls have to be made both serialized by multithreaded built-ins AND also ONLY be made from within built-ins themselves proper, and NOT from any background threads.  This all makes sense if one thinks about it for a while and maybe maps out all of the access patterns.  Yes, for those wondering, background threads can *never* call back into KSH itself since there may also be KSH itself using those same non-thread-safe calls simultaneously. So coding multithreaded within the KSH environment is indeed no cakewalk, but it is eminently doable and possible as long as one keeps their heads about them.  I (and others) have successfully created multithreaded plugins (built-ins) without problems. I have also successfully implemented many (usually dynamically created) background threads running alongside (in the background of) KSH.  But one has to be careful and mindful of what can be called at all, what might be able to be called after serialization, and what can be called from what context (either from built-in or background). Obviously I cannot really know, but it is possible that I have done the most multithreaded work within KSH as anyone has ever done, so I am only too happy to try give some pointers to others who have thought of the idea but have not yet plunged into it.

There are (yes) even many more constraints associated with multithread use, both in general and within KSH, that need to be considered for successful multithread use, but many of these are too much to go into here.  For example, although most OSes (at least the good ones) have multithread-safe access to the user-account and network database subsystems (|getpwxxx()| type calls, |gethostbyxxx()| calls, and many friends), and also that most OSes allow for multithreaded simultaneous enumeration of these kinds of calls, but *no* OS in the world today allows for simultaneous enumeration of these calls where each thread gets the whole enumeration for itself.  I had to create this sort of access myself separately, and with the help of external dynamically called on-demand daemons also under certain conditions.  The reason that this sort of thing can come into play is because the KSH thread may do an enumeration and a background thread may want to do the same (a full enumeration). Other possible failure modes are also possible. For example, if KSH tries to do an enumeration "around" a call to a built-in. But this sort of thing is a potential problem even without any threads.  So some of these things are not at all trivial or easy to get around. But as I have mentioned in previous posts, many of the issues with using multithreads within KSH also occur in other environments, and are therefore not all specific or peculiar to KSH per se.

About ASO usage, see my net post.

**Comment by DavidMorano** [comment]
@krader1961 
> There are a handful of uses of the ASO code:

My guess is that either all or most all of these includes were included unnecessarily, or something else (like maybe some possible real parallelism) is afoot.  But I very much tend to think it is the former.

I am intrigued and will look into some of this code myself to see if any of the ASO stuff is actually used.
Thanks very much.

**Comment by krader1961** [comment]
> My guess is that either all or most all of these includes were included unnecessarily, or something else (like maybe some possible real parallelism) is afoot.

Nope. Consider the `asoyield()` calls in *src/cmd/ksh93/sh/jobs.c*. These are not simply cases of including a header that isn't needed. Whether or not the `asoyield()` calls are necessary or can be replaced with `pthread_yield()` or `sched_yield()` is TBD.

>  Contrary to most belief, non-thread-safe calls can be used in multithread programs -- with the important proviso that only a single thread is using the non-thread-safe calls. 

Yes, but by definition you can't assume there is no other thread making the non-thread-safe call if your program supports more than one thread. For all `ksh` knows another thread may be calling `getpwuid()`. Yes, any plugin that creates a thread which calls a non-thread safe function is by definition broken. The whole point of such lint warnings is to advertise that the code might have unpredictable behavior given the  behavior  of  other code  linked  into the project.

**Comment by DavidMorano** [comment]
@krader1961 
> Nope. Consider the asoyield() calls in src/cmd/ksh93/sh/jobs.c.  ...

Yep. I see the call to |asoyield()|, and indeed |asoyield()| calls |sched_yield()| on those systems which have it, but there is still no indication that there are any more than one thread in the whole of KSH -- so far as I can tell.
The use of |sched_yeild()| in a program with only a single thread just becomes essentially a NOOP.

> Yes, but by definition you can't assume there is no other thread making the non-thread-safe call if your program supports more than one thread. For all ksh knows another thread may be calling getpwuid(). Yes, any plugin that creates a thread which calls a non-thread safe function is by definition broken. 

If there is only one thread within KSH itself, which currently appears to be the case (TBD), then plugins can still call non-thread-safe calls without interference with the rest of KSH (but must be serialized if more than one plugin thread attempts to do the same).  The reason being is because the rest of KSH is hung in the call to the built-in plugin itself. However, I would never advocate that any new code, being a plugin or anything else (multithreaded or not), ever call any non-thread-safe calls at any time if they can do otherwise. It was never my intention  to advocate for that. But plugins that call into either LIBAST or back into KSH itself are calling non-thread-safe subroutines. So I assume that David Korn would not have done anything to jeopardize that behavior (like by having some background thread inside KSH potentially calling subroutines that manipulate the same data structures as the non-thread-safe call from the plugin does).

So far, from all appearances, the use of ASO still appears to function as a big NOOP -- a more polite way of saying that its inclusion was unnecessary.  The inclusion might have been made in preparation for some future multithread use within KSH itself proper. but which never materialized. The real telltale is whether some real parallelism is found inside of KSH itself.  If there is, then the ASO stuff might have to stay (or be substituted out w/ the POSIX equivalents).

However, if it turns out that KSH is internally multithreaded in a non-trivial way, and if it (KSH) can call a non-thread-safe system library call from one of its background threads while (while!) a plugin is simultaneously running, then YES, it would mean that no plugin could ever call any non-thread-safe system library calls w/ guaranteed safety.  We still have to assume though that David Korn designed the whole system in a way that would still allow for plugins to call either back into KSH itself (which is generally necessary) or to call any of the LIBAST subroutines, both of which are either riddled with or entirely non-thread-safe.  We have to assume that David Korn would not jeopardize this behavior, or else his whole plugin strategy would have to be considered a faulty design. This scenario does not even have anything to do with multithreaded plugins, it would apply to even a conventional single-threaded plugin.

For myself, none of my plugin code ever calls any non-thread-safe system library calls, whether serialized or not. So even if KSH turns out to be multithreaded in a non-trivial way and indeed calls non-thread-safe system library calls from a background thread while a plugin is simultaneously running, my plugins are still correct (but only because they do not call any non-threaded system library calls).  Again, for plugins calling back into KSH or into LIBAST, Korn must have made allowance for that since many of those calls are non-thread-safe.

In the main, I think we are in complete agreement. Any plugins should only call thread-safe system library calls. For calling back into KSH or into LIBAST, there is no alternative (they being mostly non-thread-safe). We have to assume that David Korn is never doing anything in a parallel background thread within KSH that would jeopardize those calls from within a plugin from getting messed up. This seems to indicate that at best, any internal background threads within KSH probably only do some minimal function and do not do any general purpose processing which might end up calling some non-thread-safe function, either inside of itself, inside LIBAST, or in the system libraries.

My continuing guess though is that you will find that there is indeed no actual thread parallelism within KSH at this time. But even if there is, my observations above apply.

**Comment by DavidMorano** [comment]
@krader1961 
Looking through the ASO stuff, it does not in itself provide anything to create general user-visible thread parallelism. It only provides, for the most part, subroutines to manipulate data in an atomic way.  So the inclusion of the 'aso.h' header is not an indication in itself that there is any thread parallelism within KSH. But please keep looking. It is best to underhand the whole situation to better inform future plugin developers what they have to code for (the restrictions they would have to observe).

**Comment by DavidMorano** [comment]
For those following and interested: I looked through all of the AST code base and there is no indication that there is any thread parallelism at all. Certainly, if anyone finds anything different, please let me know. There are indeed calls to the ASO code (as @krader1961 pointed out). Some of the ASO usage is not exactly clear |aso_yield()|, but other ASO usage appears to be related to trying to synchronize with signal handling functions (like in 'jobs.c'). But regardless, there is still no indication of multiple thread use within KSH or libraries it calls.

For completeness I note that some of the KSH code using ASO stuff only uses it if the ASO header was found at build time. Otherwise there is other conditionally defined code to call in place of the ASO version. I am referring to the ASO code usage in 'ksh93/sh/jobs.c' and 'ksh93/sh/jobs.h'. This suggests that the ASO code is not strictly necessary, but otherwise somehow nice to have around if available.

Currently, since all of KSH and the rest of the AST libraries are not multithreaded, it should be OK to call non-thread-safe subroutines, either to system library or KSH, or to AST library, from the plug-ins. But as I have already stated. I would never advocate that any new code (plug-in or not, multithreaded o  not) ever call non-thread-safe system library calls, or any other non-thread-safe calls as much as can be avoided. For calling either back into KSH or into LIBAST (if required for some reason other than for SFIO), there is often no choice about calling non-thread-safe subroutines, since many or almost all of them are non-thread-safe. Also, if plug-ins need to call any SFIO, which they often or normally have to do, those calls are also non-thread-safe because the SFIO being actually called was already complied and linked into KSH as being non-thread-safe (again, we knew this already).

**Comment by kernigh** [comment]
Does ksh have plugins? I don't know how to make a plug-in for ksh. I suspect that src/cmd/builtin/dlls.c loads plugins, but we don't tell Meson to build that file. So we don't have plugins, with threads or not, do we?

**Comment by krader1961** [comment]
Yes, ksh supports plugins. Although the mechanism seems to be poorly documented and the last time I used ksh93u+ on a daily basis I was unaware plugins were supported. At the moment we do not have feature parity with the legacy Nmake based build. In part because as long as there are serious bugs in the core functionality (e.g., #398) that need to be fixed before anyone should even think about using this version it doesn't make sense to work on fringe features like dynamically loaded plugins.

**Comment by DavidMorano** [comment]
@kernigh 
> Does ksh have plugins?

Yes, KSH supports plugins, and has done so for some decades now (I think at least three decades).
Plugins for KSH are rather trivial to load also.  One way is to do it is via the KSH command 'builtin' as with:

$ builtin -f libfile plugin-name

Writing plugins is rather easy also with just a few caveats: they have to not leak memory nor file-descriptors (no auto cleanup like with process termination), and they must use SFIO for I/O on standard-input, standard-output, and standard-error.  Use of SFIO is required because KSH proper can pass to the plugin a memory based file rather than a file-descriptor based file. Other than these caveats, much magic is possible.  For example, using multi-threaded plugins, or starting one or more background threads to perform certain functions (like servers or other purposes).  But any magic used must be carefully managed by the plugin writer, including having the plugin dynamically unloaded at pretty much any time with responsibility of the plugin writer to free up any allocated memory structures or terminating any background threads running within the plugin.

**Comment by krader1961** [comment]
Closing since the core questions have been answered:

1) We need to assume any plugin may create threads and code accordingly.
1) We've dropped Vmalloc in favor of the platform malloc so that is no longer an issue.
1) I've removed almost all uses of ASO functionality. There are a handful left, such as `asocasptr()`, that I intend to remove in the near future.
1) It would be really nice if plugins could use stdio rather than sfio. Whether that ever happens is TBD.

See also #823 which also touches on this topic.

**Comment by DavidMorano** [comment]
@krader1961 
> It would be really nice if plugins could use stdio rather than sfio. Whether that ever happens is TBD.

I would advocate that rather than plugins being mandated to use STDIO (rather than the existing SFIO). that the plug-in I/O interface consist of just the passing of FDs just as normal processes are passed when starting up. That is: with FDs representing STDIN, STDOUT, and STDERR. In this way, the plugin is free to use whatever "add-on" I/O package they want, whether it be STDIO, SFIO, other (my favorite choice), or nothing at all! (Caveat is that packages that require their own finalizing "flush" function would have to perform that themselves before returning to the KSH monitor proper.)

Currently, plugins are required to use SFIO (at least for STDIN, STDOUT, and STDERR) because that is what KSH expects the plugin to do and is what is exploited by KSH when it passes a memory based SFIO handle to the plugin. Also, for completeness, KSH allows plugins to use its own (the AST) STDIO implementation. This is allowed because the KSH (AST) implementation of STDIO is (secretly) based on the AST SFIO package.

A future move to remove the necessity of plugins having to use SFIO (or AST STDIO in disguise) would none-the-less have to be thought through rather carefully. But I would like to see this happen eventually at some point. In the interest of full disclosure, I have suffered personally (and very substantially) by being required to use SFIO within plugins, so I am not an unbiased opinion on this. I tried to dive into the SFIO code in order to track down some bugs or feature behavior, but it was more complicated (hard to follow code) than the amount of time I wanted to spend on it, which was quite substantial.
